src/bosh_azure_cpi/lib/cloud/azure/vms/vm_manager.rb

```diff
@@ -19,23 +19,57 @@ def initialize(azure_config, registry_endpoint, disk_manager, disk_manager2, azu
-      @logger.info("create(#{bosh_vm_meta}, #{location}, #{vm_props}, ..., ...)")
+      @logger.info("create(#{bosh_vm_meta}, #{location}, #{vm_props}, #{disk_cids}, ..., ...)")
 
       instance_id = _build_instance_id(bosh_vm_meta, location, vm_props)
-
       resource_group_name = vm_props.resource_group_name
+      _check_resource_group(resource_group_name, location)
       vm_name = instance_id.vm_name
 
-      # When both availability_zone and availability_set are specified, raise an error
-      cloud_error("Only one of 'availability_zone' and 'availability_set' is allowed to be configured for the VM but you have configured both.") if !vm_props.availability_zone.nil? && !vm_props.availability_set.nil?
+      # Determine instance_type, availability_zone and availability_set
       zone = vm_props.availability_zone
-      unless zone.nil?
-        cloud_error("'#{zone}' is not a valid zone. Available zones are: #{AVAILABILITY_ZONES}") unless AVAILABILITY_ZONES.include?(zone.to_s)
-      end
+      if zone.nil?
+        availability_set_name = _get_availability_set_name(vm_props, env)
+        if vm_props.instance_type.nil?
+          calculated_instance_types = vm_props.instance_types.dup
+          @logger.debug("The cloud property 'instance_type' is nil. Will select one from '#{calculated_instance_types}', which are returned by calculate_vm_cloud_properties.")
+
+          is_persistent_disk_premium = false
+          disk_cids&.each do |disk_cid|
+            disk_id = DiskId.parse(disk_cid, @azure_config.resource_group_name)
+            disk = @disk_manager2.get_data_disk(disk_id)
+            if disk[:sku_tier] == SKU_TIER_PREMIUM
+              is_persistent_disk_premium = true
+              break
+            end
+          end
+          if is_persistent_disk_premium
+            calculated_instance_types = calculated_instance_types.select do |calculated_instance_type|
+              support_premium_storage?(calculated_instance_type)
+            end
+            @logger.debug("The disk that created VM will be attached is a Premium disk. One of the instance types '#{calculated_instance_types}' which support Premium storage will be used.")
+          end
 
-      _check_resource_group(resource_group_name, location)
+          availability_set = @azure_client.get_availability_set_by_name(resource_group_name, availability_set_name)
+          unless availability_set.nil?
+            available_vm_sizes = @azure_client.list_available_virtual_machine_sizes_by_availability_set(resource_group_name, availability_set_name)
+            available_vm_sizes = available_vm_sizes.map { |available_vm_size| available_vm_size[:name] }
+            @logger.debug("The availability set '#{availability_set_name}' supports VM sizes '#{available_vm_sizes}'")
+            calculated_instance_types = calculated_instance_types.select do |calculated_instance_type|
```

Comment:
calculated_instance_types = calculated_instance_types & available_vm_sizes

see https://stackoverflow.com/questions/5678108/how-can-i-get-the-intersection-union-and-subset-of-arrays-in-ruby

Reply:
Great suggestion.